Add Helm chart for Airflow (fix #7)

gsemet opened the pull request:
Here is a big Pull Request adding an Helm Chart for Airflow (see #7). I propose the following:
- we do the review here (I understand it can take a few time)
- I have changed some stuff in the docker files, it will requires a new image (for the moment I use mine: for testing and it works great so far `docker pull stibbons31/kube-airflow:airflow-1.8.0.0-kube-1.6`)
- once we agree, I can propose it to the official Helm repository if you agree
- I still have to add more docs, rework maybe the configuration of the templating, add a few more options (ingress or not, etc).

For information, I was greatly inspired by the [zero-to-jupyterhub helm repository](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/tree/master/jupyterhub).

**Usage:**

    helm upgrade --install --debug -f myvalues.yaml airflow /path/to/kube-airflow/airflow

**Features**
- add EditorConfig files. Because they simplify everything
- airflow.all.yaml is kept but Helm template directory is created in `./airflow`
- Propagate server config from env variables
- statup script can receive some config by ENV variable (name of rabbitmq/postgres server), at least everything  derives from the same source
- username and password for db can inherit from env
- optional `airflow.cfg` override by configmap support to set default environment variables for web, scheduler and worker (but still requires set of some env vars)
- allow ingress in endpoint such as /airflow and /flower (to do so, a [MR#2723](https://github.com/apache/incubator-airflow/pull/2723) is required to be merged first in Airflow). + Read the warning on the README about the way url prefix is to be integrated in your ingress.
- update to python 3 and debian stretch in container
- optional DAG synchronization using "git sync" to mirror a git repo for DAGs + warning about uncontrolled DAG updates in README
- load example true/false and flexible scheduler run numbers
- configurable PVC support
- use of statefulset for workers to logs can be accessed by the web server

To test this pull request:
- use image `stibbons31/kube-airflow:helm_chart_dockerfile` instead of `mumoshu/kube-airflow:1.8.0.0-1.6.1`
- `mumoshu/kube-airflow:1.8.0.0-1.6.1` needs to be updated with the new `Dockerfile.template` proposed in this PR (using more recent version of Debian and python 3, new configuration files, etc)...

Please note this PR depends on a patch on airflow ([#2723](https://github.com/apache/incubator-airflow/pull/2723)) for supporting `url_prefix` for the webservice and flower, so you **cannot** use a released version of Airflow so far. I maintain a fork with my patch until my Pull Request get integrated.

gsemet committed:
Add Helm chart for Airflow

- add EditorConfig files
- airflow.all.yaml is splited in different chart templates
- Propagate server config from env variable if defined
- statup script can receive some config by ENV variable
  (name of rabbitmq/postgres server), at least everything
  derives from the same source
- username and password can inherit from env
- fix probable buggy fernet key definition, missing enum34
  with python 2 installation
- change format of template in airflow.cfg (airflow.cfg.in)
- use configmap to set default environment variables
- parameterize number of workers
- support url prefix for flower and web ui
  allow ingress in endpoint such as /airflow and /flower
- use python 3 in docker
- optional git sync to mirror a git repo for DAGs
  + warning about uncontrolled DAG update
- debian stretch (python >3.4)
- load example true/false and flexible scheduler run nums

lsowen commented:
@Stibbons a helm chart would be awesome.  However, do you think it might be a better fit for https://github.com/kubernetes/charts?  That would allow better discoverability, review, and long term maintenance.

gsemet commented:
Yes, I plan to land it to kube-airflow for testing, and if it works, merge it to kubernetes/charts

chhetripradeep commented on airflow/templates/configmaps.yaml:
```diff
@@ -0,0 +1,28 @@
+  {{- if .Values.airflow.fernet_key }}
+  FERNET_KEY: "{{- .Values.airflow.fernet_key -}}"
+  {{- end }}
+  RABBITMQ_HOST: "{{- .Values.airflow.prefix -}}{{- .Values.db.rabbitmq.basename -}}"
```
Typo: `hostname` instead of `basename`

gsemet replied:
I wanted to emphasis it is not a full hostname. the final hostname will be prefix with airflow.prefix

chhetripradeep replied:
In that case shouldn't the corresponding variable in `values.yaml` be using `basename` instead of `hostname`

gsemet replied:
hum this part works pretty well, I don't see your point. values.yaml allows user to specify the basehostname, and here it prefixed with the generic prefix (I use it so it can be possible to start several airflow instance in the same namespace)

chhetripradeep commented on airflow/templates/configmaps.yaml:
```diff
@@ -0,0 +1,28 @@
+  {{- if .Values.db.rabbitmq.user }}
+  RABBITMQ_CREDS: "{{- .Values.db.rabbitmq.user -}}:{{- .Values.db.rabbitmq.password -}}"
+  {{- end }}
+  POSTGRES_HOST: "{{- .Values.airflow.prefix -}}{{- .Values.db.postgres.basename -}}"
```
Typo: `hostname` instead of `basename`

chhetripradeep commented:
@Stibbons I tried your helm chart with your docker image - few things didn't work for me.

- scheduler & worker container - i got the error saying: ```entrypoint.sh: python3 not found```
- flower container failed with error - ```airflow: error: argument subcommand: invalid choice: '/usr/local/airflow/flowerconfig.py' (choose from 'resetdb', 'render', 'variables', 'connections', 'pause', 'task_failed_deps', 'version', 'trigger_dag', 'initdb', 'test', 'unpause', 'dag_state', 'run', 'list_tasks', 'backfill', 'list_dags', 'kerberos', 'worker', 'webserver', 'flower', 'scheduler', 'task_state', 'pool', 'serve_logs', 'clear', 'upgradedb')```

gsemet commented:
You need to use `stibbons31/kube-airflow:airflow-HEAD-patched-kube-1.6` instead of `mumoshu/kube-airflow:1.8.0.0-1.6.1` until `mumoshu/kube-airflow:1.8.0.0-1.6.1` is updated with the new `Dockerfile.template` (using more recent version of Debian and python 3)...

PS: I have fixed an issue today can you try again by retrieving the latest version of this PR

gsemet commented:
Also, I will probably propose a version that store the complete airflow.cfg in a configmap, it should be much simpler

chhetripradeep commented on airflow/values.yaml:
```diff
@@ -0,0 +1,53 @@
+    use_embedded: true
+    user: airflow
+    password: airflow
+    hostname: rabbitmq
```
I mean shouldn't this key be `basename` instead of `hostname`.

gsemet replied:
ho, yes indeed

chhetripradeep commented:
@Stibbons I used your image `stibbons31/kube-airflow:airflow-HEAD-patched-kube-1.6` and now i am getting this exception: 

```[2017-10-25 18:23:45,732] {driver.py:123} INFO - Generating grammar tables from /usr/lib/python3.5/lib2to3/PatternGrammar.txt
Traceback (most recent call last):
  File "/usr/local/bin/airflow", line 17, in <module>
    from airflow import configuration
  File "/usr/local/lib/python3.5/dist-packages/airflow/__init__.py", line 31, in <module>
    from airflow.models import DAG
  File "/usr/local/lib/python3.5/dist-packages/airflow/models.py", line 138, in <module>
    class DagBag(BaseDagBag, LoggingMixin):
  File "/usr/local/lib/python3.5/dist-packages/airflow/models.py", line 164, in DagBag
    include_examples=configuration.getboolean('core', 'LOAD_EXAMPLES')):
  File "/usr/local/lib/python3.5/dist-packages/airflow/configuration.py", line 804, in getboolean
    return conf.getboolean(section, key)
  File "/usr/local/lib/python3.5/dist-packages/airflow/configuration.py", line 626, in getboolean
    'boolean (received "{}").'.format(section, key, val))
airflow.exceptions.AirflowConfigException: The value for configuration option "core:LOAD_EXAMPLES" is not a boolean (received "").

gsemet commented:
@chhetripradeep should be fixed on latest rev.

chhetripradeep commented:
@Stibbons sorry the above exception is my bad. i didnt pull the latest helm chart branch.

gsemet commented:
I just fixed it now :)

chhetripradeep commented:
@Stibbons Ack that the chart now works like a charm 🎉  🎉

chhetripradeep commented:
I am also new to helm and k8s ecosystem but few feedbacks i would suggest - 

- Using PVC for databases.
- Storing complete airflow.cfg in a configmap.
- Adding a flag to enable/disable ingress, Based on that we can disable/enable nodePort for service.

Overall, chart looks great.

gsemet commented:
There will be probably 2 helm charts:
- 1 for kubernetes celery (this one) without my url_prefix, no configmap, maybe git-sync and python3
- 1 for kubernetes executor (the patch is still in the queue in airflow), probably in airflow 2, with configmap, pvc, git-sync as well, ...

I'll add quickly the ingress switch, i'll try pvc if i have the time, and i also wanted to switch to config map but with this image and the entrypoint.sh it will be tricky to do

grantnicholas reviewed (commented):
Just a note. Be careful with git-sync and airflow. If the scheduler reloads a dag in the middle of a dagrun then the dagrun will run actually start using the new version of the dag in the middle of execution. This is a known issue with airflow and it means it's unsafe in general to use a git-sync like solution with airflow without:
1. using explicit locking, ie never pull down a new dag if a dagrun is in progress
2. make dags immutable, never modify your dag always make a new one

gsemet commented:
Indeed, good to know. It is just an option, and probably needs more documentation. By the way, do you know if there are official "recommendation" on airflow website, about Dag update/deployment? I don't see such information in the doc

gsemet commented:
with your permission, I added your recommendations in the readme and in config.yaml file. Always good to know

gsemet committed:
requirements*.txt support + docs

Signed-off-by: Gaetan Semet <gaetan@xeberon.net>

grantnicholas commented:
@Stibbons sure go ahead. I talked with the airflow devs about this as part of the kubernetes executor stuff. They acknowledged it's an issue and mentioned several workarounds. I totally agree with you documenting the "suggested" way for updating dags would be super useful. I can ask around the airflow mailing list and see what others are doing if that's helpful.

gsemet commented:
Yes I'll start a thread on the mailing list about this subject. I have modified this chart today to support two DAGs deployment mechanisms:
- git-sync (with the warning you raised)
- embed DAGs inside the docker (require custom build of the docker images)
And a proposal of support for dependency management (requirements/*.txt). I think it could allow to easily adapt our DAGs with python module (ex: scikit learn,...).

Can you review the README I propose, see my proposal seems viable ?
Link to [rendered README](https://github.com/Stibbons/kube-airflow/blob/helm_chart/README.md#dags-deployment-embedded-dags-or-git-sync)

Commit:
This commit adds and icon to the Chart.
This is not required but still shows up in helm lint tests

Commit:
This commit adds a helpers to access name variables in the templates

Commit:
Adding of optional pvc WIP
This commit adds an optional PVC for the dags directory.

Commit:
Change the PVC access mode
Ths changes the acces mode to ReadWriteMany because it will be accessed
more than one pod

# Conflicts:
#	airflow/values.yaml

Commit:
This commit adds a more up to date version
of the pvc and values to go with the persistence section

# Conflicts:
#	airflow/templates/pvc.yaml
#	airflow/values.yaml

gsemet committed:
Refactor deployments and add a dags path variable

This commit breaks the deployments out into separate files.
The purpose of this is for troubleshooting when making
changes to the deployment and fixing errors when finding
which deployment and line number is causing a problem.

This change also make the dag_path which is used for
volumeMounts within deployments.

gsemet commented:
@Sedward, please check this version if I did the rebase over my changes correctly. This adds PVC support for Airflow Chart. I'll test it more intensively tomorrow

gsemet committed:
Fixes after PVC merge

gsemet committed:
Protect all names

Signed-off-by: Gaetan Semet <gaetan@xeberon.net>

gsemet committed:
fix some issues on the charts

gsemet committed:
fix in airflow.cfg

and remove '.in'. It will allow to override the
configuration file by configmap more easily

gsemet committed:
airflow.cfg configmap

gsemet committed:
beware of the sed when replacing url_prefix

do not use sed s/../../
but sed s##..#...

gsemet committed:
update README

nathanbreitsch commented on script/entrypoint.sh:
```diff
@@ -1,20 +1,42 @@
-sed -i "s/{FERNET_KEY}/${FERNET_KEY}/" $AIRFLOW_HOME/airflow.cfg
+sed -i "s/{{ FERNET_KEY }}/${FERNET_KEY}/" $AIRFLOW_HOME/airflow.cfg
+sed -i "s/{{ POSTGRES_HOST }}/${POSTGRES_HOST}/" $AIRFLOW_HOME/airflow.cfg
+sed -i "s/{{ POSTGRES_CREDS }}/${RABBITMQ_CREDS}/" $AIRFLOW_HOME/airflow.cfg
```
This looks like a typo.  It breaks when `POSTGRES_CREDS` != `RABBITMQ_CREDS`.

gsemet replied:
ho yes

gsemet replied:
should be fixed now

gsemet committed:
Fix credentials applying + update config.yaml example

gsemet committed:
Minor glitch fixes

gsemet committed:
make helm-check

gsemet commented:
Today's changes brings airflow.cfg configmap support !

chhetripradeep commented:
@Stibbons May i know which docker image i can test it with. I tried with stibbons31/kube-airflow:airflow-kube_executor_url_prefix-kube-1.6-v8 but i see errors in `web` pod saying "ImportError: No module named 'kubernetes'"

gsemet commented:
You can try with image: `stibbons31/kube-airflow:helm_chart_dockerfile`

Commit:
Rework

- statefulset worker (allow access to port and so logs)
- different ingress annotations between flower and webserver
- Add /usr/bin/python symblink in Dockerfile
- ensure file overwrite with git sync
- using rolling update statefulset
- allow to pickle DAG from scheduler to worker
  so git sync on worker MIGHT be useless)
- update README
- set scheduler num runs to 1
  will restart scheduler at each execution

Signed-off-by: Gaetan Semet <gaetan@xeberon.net>

gsemet commented:
I am pretty happy about the new change I added. Airflow is now functionnal in Kuberneter
- workers logs can be retreived by the webserver (thanks to statefulset)
- can mount application to http://airflow.mycompany.com/celery/admin (for webserver) and http://airflow.mycompany.com/celery/flower (for Flower)

gsemet commented:
Update:
- statefulset worker (allow access to port and so logs)
- different ingress annotations between flower and webserver
- Add /usr/bin/python symblink in Dockerfile
- ensure file overwrite with git sync
- allow to pickle DAG from scheduler to worker
  (so git sync on worker MIGHT be useless)
- update README
- set scheduler num runs to 1
  will restart scheduler at each execution

gsemet committed:
update about values.yaml

gsemet committed:
Values configuration

gsemet committed:
statefulset: burst mode

gsemet committed:
Cleaner example

gsemet committed:
Rolling update

gsemet committed:
Add web service liveness probe

Commit:
no wait

jamescarr commented:
How is this currently going? My team has some work coming up soon to "kuberneterize" our data team's airflow setup and my first instinct was to write a helm chart. 

More than happy to lend a hand if we can!

gsemet commented:
I actually maintain my fork of [kube-airflow](https://github.com/Stibbons/kube-airflow). I try to keep this pr the most up to date but i get no review from the original author. I have diverged quite a lot from `airflow.all.yaml`, but at least most features works with celery executor

bbhoss commented:
Happy new year. Just chiming in along the same lines as @jamescarr. I am surprised this doesn't already exist, and would be happy to help. Has there been any response from the maintainers at all?

gsemet commented:
Note:
- I am planning to officially add this chart in the official Helm repository
- Just need an evolution to support the url prefix change that has been integrated in airflow.

jpds committed:
Removed embedded postgresql pieces.

jpds committed:
Added dependency on postgresql chart.

jpds committed:
Leverage postgres dependency in configmaps.

jpds committed:
airflow/values.yaml: Removed concourse reference.

gsemet committed:
Merge pull request #13 from jpds/postgresql-dependency

Added a dependency on main PostgreSQL chart

jpds committed:
Use puckel image and split out imageTag.

jpds committed:
Removed airflow_cfg and all references to it.

jpds committed:
Use database environment variables for all services.

jpds committed:
Integrated redis chart and variables.

jpds committed:
deployments-web.yaml: Fixed health check endpoint.

jpds committed:
values.yaml: Removed airflow_cfg.

jpds committed:
Removed RabbitMQ references.

jpds committed:
services.yaml: Don't use NodePorts to allow for multiple deployments on
same cluster.

jpds committed:
templates/configmaps-env.yaml: Removed.

jpds committed:
Adapted deployment labels to `helm create` defaults.

jpds committed:
Replaced prefix option with dynamically generated helm release name.

jpds committed:
Added airflow.config option for dynamic airflow variable setting.

jpds committed:
airflow/values.yaml: config value for dynamic airflow env assignment.

jpds committed:
statefulsets-workers: Specify imageTag.

jpds committed:
statefulsets-workers: Fixed name field for lint.

jpds committed:
services.yaml: Match ingress serviceNames.

jpds committed:
ingresses.yml: Match other chart label annotations and use subdomain for
flower instead of path prefixes.

jpds committed:
values.yaml: Removed rabbitmq section.

gsemet committed:
install airflow HEAD now my patch has been integrated

gsemet committed:
Merge pull request #14 from jpds/puckel-image

Use puckel/docker-airflow image

mumoshu commented:
@gsemet Hey! I finally managed to catch all the things up made until now.

Thanks for the all the your efforts and the fork, and the PR to official Kubernetes charts repository 👍 

Everyone, see his awesome work at https://github.com/kubernetes/charts/pull/3959.

I'll keep this repository for mostly for a reference for a while.
But I believe - the general direction should be treat @gsemet's work as the primary, and contribute any missing parts(if any) to his k8s chart and eventually obsolete this as superseded by it.
